### PR TITLE
FIX #127:Remove MM_ScanClassesMode and dependencies from OMR

### LINKED ISSUE #127: Remove MM_ScanClassesMode and dependencies from OMR
MM_ScanClassesMode tracks state changes in a Java-side process (collecting and scanning class-related roots and dependencies for ConcurrentGC) and should be removed to Java source code repo. Additional CLI methods must be defined to permit ConcurrentGC to interact with Java threads similarly as with other language threads, and Java's MM_ScanClassesMode and related dependencies must be moved into respective Java CLI method implementations. 

This is the next logical step toward a language-agnostic implementation of ConcurrentGC. Other changes are expected after this step is complete, but further work will be completed in new issues.

### PR DESCRIPTION
**Pull request by ktbriggs**
Complete ConcurrentGC integration with example glue and GCTest.

Implement VM access functions for example/glue. Remove redundant and
unused MM_EnvironmentBase::_exclusiveCount member; use instead
OMR_VMThread::exclusiveCount as previously and extensively used.

Remove updates to OMR_VMThread::exclusiveCount in MM_EnvironmentBase.
This field must be maintained by the client language, or existing client
language (Java) must be refactored to allow maintenance of this field in
MM_EnvironmentBase. 

Fix dead object detection for example/glue (GC_ObjectModel).

Add MM_ObjectModel::initializeObject() to allow omrgcalloc 
methods to set new object header before paying allocation tax (this 
is a stop-gap measure for now -- updates to omrgcalloc methods will
follow in separate task).

Add MM_CollectorLanguageInterface::updateWriteBarrier(), to be used
in contexts where another agent has already effected assignment of 
child object to parent slot.

Signed-off-by: Kim Briggs <briggs@ca.ibm.com>

### CODE REVIEW

**Comment by ktbriggs** [comment]
This commit is follow-up to previous work on ConcurrentGC (f4f0bff). Main task was to implement exclusive/non-exclusive VM access and save/restoreObject() for example/glue so that concurrent helper thread could operate in fvtest (omrgctest). Previous commit also disabled payAllocationTax(), it is enabled in this commit. Other changes in this commit relate to fixes for issues found while developing this (dead object detection), extension of default write barrier, and some cleanup.

The changes to MM_EnvironmentBase VM access methods impact Java as well. I have a clean all-platform personal build for Java (sanity+continuous tests) with this commit in on the farm. Also tested gencon and optavgpause with SPECjvm2008 on rhelvm46 (24 CPUs). Seems to be fine. 

@dmitripivkine please review for submission to eclipse/omr. will ask @charliegracie to approve and accept pull request when you have had a chance to review and comment.

**Comment by ktbriggs** [comment]
There is a problem in MM_AllocateInitialization::initializeAllocation(..). The setFlags() call should be like this:

```
setFlags(objectPtr, OMR_OBJECT_METADATA_FLAGS_MASK, 0)
```

The current call with 0 as 2nd argument does not clear existing bits in the flags field of the object header. I'm holding onto a commit to address review comments and will push it to my fork when all comments are in.

File: gc/base/standard/Scavenger.cpp [hunk 1]
```diff
@@ -3011,11 +3011,12 @@ MM_Scavenger::backOutFixSlot(GC_SlotObject *slotObject)
 		MM_ForwardedHeader forwardHeader(objectPtr, OMR_OBJECT_METADATA_SLOT_OFFSET);
 		Assert_MM_false(forwardHeader.isForwardedPointer());
 		if (forwardHeader.isReverseForwardedPointer()) {
-			omrobjectptr_t fwdObjectPtr = forwardHeader.getReverseForwardedPointer();
+			slotObject->writeReferenceToSlot(forwardHeader.getReverseForwardedPointer());
 #if defined(OMR_SCAVENGER_TRACE_BACKOUT)
+			OMRPORT_ACCESS_FROM_OMRVM(_omrVM);
+			omrtty_printf("{SCAV: Back out object slot %p[%p->%p]}\n", slotPtr, objectPtr, *slotPtr);
 			Assert_MM_true(isObjectInEvacuateMemory(fwdObjectPtr));
```
Comment (charliegracie) [inline gc/base/standard/Scavenger.cpp]: I believe there is a compile failure here if you compile with the debug code. fwdObjectPtr is no longer declared.

File: gc/startup/omrgcalloc.cpp [hunk 2]
```diff
@@ -60,18 +60,31 @@ allocHelper(OMR_VMThread * omrVMThread, size_t sizeInBytes, uintptr_t flags, boo
...
 
-	/* OMRTODO: Should we use zero TLH instead of memset? */
 	if (NULL != heapBytes) {
 		if (J9_ARE_ALL_BITS_SET(flags, OMR_GC_ALLOCATE_ZERO_MEMORY)) {
-			uintptr_t size = allocdescription.getBytesRequested();
-			memset(heapBytes, 0, size);
+			memset(heapBytes, 0, allocdescription.getBytesRequested());
 		}
+		/* TODO: Provide a generic object initialization model interface to be implemented for each language context. */
+#if defined(OMR_EXAMPLE)
```
Comment (charliegracie) [inline gc/startup/omrgcalloc.cpp]: This should not be here as this is technically language code.

File: include_core/omr.h [hunk 3]
```diff
@@ -177,6 +177,8 @@ typedef struct OMR_VMThread {
 	struct movedObjectHashCode movedObjectHashCodeCache;
 
 	int32_t _attachCount;
+
+	void *savedObject; /**< holds new object allocation until object can be attached to reference graph (see MM_AllocationDescription::save/restoreObjects()) */
```
Comment (charliegracie) [inline include_core/omr.h]: I know not all fields have the _ but they should.  Please add the underscore for this field please.

**Comment by amicic** [comment]
I think the question was would it compile with the Scavenger build flag disabled.

**Comment by ktbriggs** [comment]
That too :). Turning on all of those build flags generates a large volume of output even for gctest so I was reluctant to push a debug build to github/travis to verify (I don't know how to set up a compile only/no test build for travis). So I verified locally with manual build from local repo with all Scavenger build flags enabled. Compiles and runs fine.

**Comment by ktbriggs** [comment]
I did miss one of your (Aleks) issues -- checking VM exclusive access in concurrent helper thread. This fix is in most recent commit (b2d9092).

**Comment by ktbriggs** [comment]
@amicic @charliegracie any comments on this as is? If not I'll squash latest series of commits and push result back for pull into eclipse/master.

**Comment by amicic** [comment]
Change in getConHelperRequest will have   _conHelpersActivationMonitor being exercised more often. In practice may not be an issue, since isExclusiveAccessRequestWaiting is probably more often false. 

This also affects the logic a bit for the first getConHelperRequest  invocation site - i'm not completely convinced it's same as before. Although it may have positive effects, like detecting that we should/could block the concurrent thread(s) a bit earlier than before.

Overall, I believe it's ok, but I still wonder if that shuffling is worth the risk.
----
I'd make acquire/release Exclusive in base ELI class pure virtual. The example in ELIImpl is sufficient.

Other than that looks ok.

**Pull request closed without merging**